File: src/Application.Resources/scripts/event-lua/MK_PrimeMinister.lua
local BaseEvent = require("scripts/event-lua/__BasePQ")

-- 配置事件参数
local config = {
    name = "MK_PrimeMinister",
    instanceName = "MK_PrimeMinister_",
    minPlayers = 1,
    maxPlayers = 3,
    minLevel = 30,
    maxLevel = 255,
    entryMap = 106021600,
    entryPortal = 1,
    exitMap = 106021402,
    recruitMap = 106021402,
    clearMap = 0,
    warpTeamWhenClear = false,
    minMapId = 106021600,
    maxMapId = 106021600,
    eventTime = 10,
    maxLobbies = 1,

    bossId = 3300008
}


-- 创建自定义事件
local Sample = BaseEvent:extend()

-- 没办法通过设置处理的就在这里进行重载
function Sample:setup(level, lobbyid)
    local eim = BaseEvent.setup(self, level, lobbyid)
    -- xxx
    self:respawn(eim)
    return eim
end

function Sample:primeMinisterCheck(eim)
    local map = eim:getMapInstance(self.entryMap);

    for _, player in ipairs(map:getAllPlayers()) do
        if (player:getQuestStatus(2333) == 1 and player:getAbstractPlayerInteraction():getQuestProgressInt(2333, self.mobId) == 0) then
            return true;
        end
    end

    return false;
end

function Sample:respawn(eim)
    if (self:primeMinisterCheck(eim)) then
        eim:startEventTimer(self.eventTime);

        local weddinghall = eim:getMapInstance(self.entryMap);
        weddinghall:getPortal(self.entryPortal):setPortalState(false);
        weddinghall:spawnMonsterOnGroundBelow(LifeFactory:getMonster(self.mobId), Point(292, 143));
    else
        eim:schedule("respawn", 10000);
    end
end

function Sample:monsterKilled(mob, eim)
    if (mob:getId() == self.mobId) then
        eim:getMapInstance(self.entryMap):getPortal(self.entryPortal):setPortalState(true);

        eim:showClearEffect();
        eim:clearPQ();
    end
end

Sample:new(config)

File: src/Application.Plugin.Script/Events/MK_PrimeMinister.cs
using Application.Core.Channel.Net.Packets;
using Application.Core.Game.Life;
using Application.Core.Game.Maps;
using Application.Core.Game.Players;
using Application.Core.scripting.Events.Instances;
using Application.Core.scripting.Events.Templates;
using CommunityToolkit.HighPerformance.Helpers;
using server.life;
using System.Drawing;

namespace Application.Plugin.Script.Events
{
    internal class MK_PrimeMinister : AbstractPartyQuestEventTemplate
    {
        public MK_PrimeMinister() : base(nameof(MK_PrimeMinister))
        {
            EventTime = 10 * 60;
            MinCount = 1;
            MaxCount = 3;
            MinLevel = 30;
            MaxLevel = 255;

            EntryMap = 106021600;
            EntryPortal = 1;
            ExitMap = 106021402;
            MinMap = 106021600;
            MaxMap = 106021600;
            RecruitMap = 106021402;

        }

        public override async Task OnSetup(AbstractEventInstanceManager eim, int level, int lobbyId)
        {
            await base.OnSetup(eim, level, lobbyId);

            var weddinghall = await eim.getMapInstance(EntryMap);
            weddinghall.getPortal(1)?.setPortalState(false);
        }

        public override async Task OnPlayerEntry(AbstractEventInstanceManager eim, Player chr)
        {
            await base.OnPlayerEntry(eim, chr);

            if (chr.Id == eim.getLeaderId())
            {
                var questStatus = chr.GetOrAddQuest(2333);
                if (questStatus.getStatus() != client.QuestStatus.Status.NOT_STARTED)
                {
                    await SpawnBoss(eim);
                }
            }
        }

        int mobId = 3300008;
        public override async Task OnMobKilled(AbstractEventInstanceManager eim, Monster mob, ICombatantObject? killer)
        {
            if (mob.getId() == mobId)
            {
                var map = await eim.getMapInstance(EntryMap);
                map.getPortal(1)?.setPortalState(true);

                await eim.showClearEffect();
                await eim.clearPQ();
            }
        }

        public async Task SpawnBoss(AbstractEventInstanceManager eim)
        {
            var weddinghall = await eim.getMapInstance(EntryMap);
            var pos = new Point(292, 143);
            await weddinghall.spawnMonsterOnGroundBelow(LifeFactory.Instance.getMonster(mobId), pos);
            await weddinghall.broadcastMessage(FieldEffectPacket.Summon(23, pos.X, pos.Y)); // name=23一样？
        }
    }
}
